pkg/sink/sink.go

```diff
@@ -0,0 +1,130 @@
+		return
+	}
+
+	event, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		log.Printf("Error reading event body: %s", err)
+		return
+	}
+
+	for _, tbRef := range el.Spec.TriggerBindingRefs {
+		tbNamespace := tbRef.Namespace
+		if tbNamespace == "" {
+			tbNamespace = r.EventListenerNamespace
+		}
+		tb, err := r.TriggersClient.TektonV1alpha1().TriggerBindings(tbNamespace).Get(tbRef.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Printf("Error getting TriggerBinding %s in Namespace %s: %s", tbRef.Name, tbNamespace, err)
+			continue
+		}
+
+		for _, templateBinding := range tb.Spec.TemplateBindings {
+			ttNamespace := templateBinding.TemplateRef.Namespace
+			if ttNamespace == "" {
+				ttNamespace = r.EventListenerNamespace
+			}
+			tt, err := r.TriggersClient.TektonV1alpha1().TriggerTemplates(ttNamespace).Get(templateBinding.TemplateRef.Name, metav1.GetOptions{})
+			if err != nil {
+				log.Printf("Error getting TriggerTemplate %s in Namespace %s: %s", templateBinding.TemplateRef.Name, ttNamespace, err)
+				continue
+			}
+
+			params, err := template.ApplyEventToParams(event, templateBinding.Params)
+			if err != nil {
+				log.Printf("Error applying event to TriggerBinding params: %s", err)
+				continue
+			}
+			params = template.AddDefaultParamsFromSpecs(params, tt.Spec.Params)
+
+			for _, rt := range tt.Spec.ResourceTemplates {
+				rtRaw := template.ApplyParamsToResourceTemplate(params, rt.RawMessage)
+				err = createResource(r.KubeClient, rtRaw)
+				if err != nil {
+					log.Printf("Error creating resources from resource template %s: %s", string(rtRaw), err)
+					continue
+				}
+				log.Printf("Successfully created resource: %s", string(rtRaw))
+			}
+		}
+	}
+}
+
+// createResource uses the kubeClient to create the resource defined in the
+// TriggerResourceTemplate and returns any errors with this process
+func createResource(kubeClient *kubeclientset.Clientset, rt json.RawMessage) error {
+	// Assume the TriggerResourceTemplate is valid (it has an apiVersion and Kind)
+	result := kubeClient.RESTClient().Post().
+		RequestURI(getRequestURI(rt)).
+		Body([]byte(rt)).
+		SetHeader("Content-Type", "application/json").
+		Do()
+	if result.Error() != nil {
+		return result.Error()
+	}
+	return nil
+}
+
+// getRequestURI returns the URI for a post request to the kubernetes API
+// REST endpoint to create the TriggerResourceTemplate
+func getRequestURI(rt json.RawMessage) string {
+	// Assume the TriggerResourceTemplate is valid (it has an apiVersion and Kind)
+	apiVersion := gjson.GetBytes(rt, "apiVersion").String()
+	kind := gjson.GetBytes(rt, "kind").String()
+	namespace := gjson.GetBytes(rt, "metadata.namespace").String()
+	var uri string
+	if apiVersion == "v1" {
+		uri = "api/v1"
+	} else {
+		uri = path.Join(uri, "apis", apiVersion)
+	}
+	if namespace != "" {
+		uri = path.Join(uri, "namespaces", namespace)
+	}
+	uri = path.Join(uri, fmt.Sprintf("%ss", strings.ToLower(kind)))
```

Comment:
I wonder if the resources we are created could/would have some plural form where we shouldn't just assume the plural just adds an 's' like we do here: https://github.com/tektoncd/pipeline/blob/master/config/300-clustertask.yaml#L21. Maybe we could make a request to obtain the plural form?

Reply:
True, I wonder how kubernetes does it. I assume I can make a request to the kube client to get the plural name of a resource, although I'm not sure what I would query. I'll look into this, but does anyone know off the top of their head?